File: packages/ios-im-ui/Sources/FlareIMUI/Models/LocateHighlight.swift
import CoreGraphics
import Foundation
import SwiftUI

/// The mark a located row wears while the reader's eye catches up with the jump.
struct FlareLocateHighlightValue: Equatable {
    /// Whether the row is marked at all. False once the window is over.
    let marked: Bool
    /// The ring's opacity.
    let alpha: Double
    /// How far the ring reaches beyond the bubble's edge, in points.
    let spread: CGFloat

    static let none = FlareLocateHighlightValue(marked: false, alpha: 0, spread: 0)
}

/// The one locate-mark rule, shared with the Flutter and Compose kits and with Vue's stylesheet, and tested
/// against `spec/locate-highlight-vectors.json`.
enum FlareLocateHighlight {
    /// How long a located row stays marked. One number: the ring reaches nothing exactly as the window closes.
    static let durationMs: Double = 1600

    private static let stops: [(atMs: Double, alpha: Double, spread: CGFloat)] = [
        (0, 0.36, 0),
        (700, 0.18, 10),
        (durationMs, 0, 18),
    ]

    /// Held still for a reader who asked for less motion: the middle stop, for the whole window. Less motion
    /// is not no answer — the jump still has to say where it landed.
    private static let reducedMotion = FlareLocateHighlightValue(marked: true, alpha: 0.18, spread: 10)

    /// `elapsedMs` counts from the moment the list started moving towards the row.
    static func resolve(elapsedMs: Double, reduceMotion: Bool) -> FlareLocateHighlightValue {
        guard elapsedMs < durationMs else { return .none }
        guard !reduceMotion else { return reducedMotion }
        let elapsed = max(0, elapsedMs)
        for index in 0..<(stops.count - 1) {
            let from = stops[index]
            let to = stops[index + 1]
            if elapsed > to.atMs { continue }
            let t = (elapsed - from.atMs) / (to.atMs - from.atMs)
            return FlareLocateHighlightValue(
                marked: true,
                alpha: from.alpha + (to.alpha - from.alpha) * t,
                spread: from.spread + (to.spread - from.spread) * t
            )
        }
        return .none
    }
}

/// Which row is marked after a jump, and when its window started. The list owns both and puts them in the
/// environment; a bubble takes the mark only when it is the marked one. One row at a time: a second jump
/// replaces this value rather than adding to it.
struct FlareLocateMark: Equatable {
    let messageId: String?
    let startedAt: Date

    static let none = FlareLocateMark(messageId: nil, startedAt: .distantPast)
}

private struct FlareLocateMarkKey: EnvironmentKey {
    static let defaultValue = FlareLocateMark.none
}

extension EnvironmentValues {
    var flareLocateMark: FlareLocateMark {
        get { self[FlareLocateMarkKey.self] }
        set { self[FlareLocateMarkKey.self] = newValue }
    }
}

File: packages/flutter-im-ui/lib/src/models/locate_highlight.dart
/// The mark a located row wears while the reader's eye catches up with the jump.
class FlareLocateHighlight {
  const FlareLocateHighlight({
    required this.marked,
    required this.alpha,
    required this.spread,
  });

  /// Whether the row is marked at all. False once the window is over.
  final bool marked;

  /// The ring's opacity.
  final double alpha;

  /// How far the ring reaches beyond the bubble's edge, in logical pixels.
  final double spread;

  static const FlareLocateHighlight none = FlareLocateHighlight(
    marked: false,
    alpha: 0,
    spread: 0,
  );
}

/// How long a located row stays marked. One number: the ring reaches nothing
/// exactly as the window closes.
const int flareLocateHighlightDurationMs = 1600;

const _stops = <({int atMs, double alpha, double spread})>[
  (atMs: 0, alpha: 0.36, spread: 0),
  (atMs: 700, alpha: 0.18, spread: 10),
  (atMs: flareLocateHighlightDurationMs, alpha: 0, spread: 18),
];

/// Held still for a reader who asked for less motion: the middle stop, for the
/// whole window. Less motion is not no answer — the jump still has to say where
/// it landed.
const _reducedMotion = FlareLocateHighlight(marked: true, alpha: 0.18, spread: 10);

/// The one locate-mark rule, shared with the SwiftUI and Compose kits and with
/// Vue's stylesheet, and tested against `spec/locate-highlight-vectors.json`.
/// [elapsedMs] counts from the moment the list started moving towards the row.
FlareLocateHighlight flareLocateHighlight(
  double elapsedMs, {
  required bool reduceMotion,
}) {
  if (elapsedMs >= flareLocateHighlightDurationMs) {
    return FlareLocateHighlight.none;
  }
  if (reduceMotion) return _reducedMotion;
  final elapsed = elapsedMs < 0 ? 0.0 : elapsedMs;
  for (var i = 0; i < _stops.length - 1; i++) {
    final from = _stops[i];
    final to = _stops[i + 1];
    if (elapsed > to.atMs) continue;
    final t = (elapsed - from.atMs) / (to.atMs - from.atMs);
    return FlareLocateHighlight(
      marked: true,
      alpha: from.alpha + (to.alpha - from.alpha) * t,
      spread: from.spread + (to.spread - from.spread) * t,
    );
  }
  return FlareLocateHighlight.none;
}
